### PR TITLE
Data.NameMap: remove runtime check on HOF name changes

### LINKED ISSUE #14901: `daml damlc validate-dar` fails in the presence of interface instances
Given

```yaml
# daml.yaml
sdk-version: 2.4.0-snapshot.20220830.10494.0.4622de48
build-options: [--target=1.15]
name: main
source: Main.daml
version: 0.0.1
dependencies:
  - daml-prim
  - daml-stdlib
  - daml-script
```

```haskell
-- Main.daml
module Main where

interface I where
  viewtype T

template T
  with
    p : Party
    i : Int
  where
    signatory p
    interface instance I for T where
      view = this
```

```bash
> daml build && daml damlc validate-dar .daml/dist/main-0.0.1.dar
2022-09-01 12:54:36.32 [INFO]  [build] 
Compiling main to a DAR.

2022-09-01 12:54:36.80 [INFO]  [build] 
Created .daml/dist/main-0.0.1.dar
damlc: Data.NameMap.traverse: function changed name from Qualified {qualPackage = PRImport (PackageId {unPackageId = "86487a56ad17c2c79d24de6bb210df237804721949c7b27d61468bf3afdd18ba"}), qualModule = ModuleName {unModuleName = ["Main"]}, qualObject = TypeConName {unTypeConName = ["I"]}} to Qualified {qualPackage = PRSelf, qualModule = ModuleName {unModuleName = ["Main"]}, qualObject = TypeConName {unTypeConName = ["I"]}}
CallStack (from HasCallStack):
  error, called at libs-haskell/da-hs-base/src/Data/NameMap.hs:132:21 in libs-haskellZSda-hs-baseZSda-hs-base:Data.NameMap
```

The problem is that `TemplateImplements` and `InterfaceCoImplements` are stored in `NameMap`s. `validateDar` tries to rename all references to the current package ([src](https://github.com/digital-asset/daml/blob/main/compiler/damlc/daml-compiler/src/DA/Daml/Compiler/Validate.hs#L56-L58)), but that fails because `NameMap`'s `traverse` fails if the function changes the `Name`s

### PR DESCRIPTION
**Pull request by akrmn**
Fixes #14901

### CODE REVIEW

File: libs-haskell/da-hs-base/src/Data/NameMap.hs [hunk 1]
```diff
@@ -133,6 +133,13 @@ traverse f (NameMap ras _) = build <$> Prelude.traverse f' (reverse ras)
           "Data.NameMap.traverse: function changed name from " ++ show n ++ " to " ++ show (name y)
     build as = NameMap (reverse as) (HMS.fromList as)
 
+-- | @traverseAsList f m@ applies the function @f@ to all elements of @m@ in the
+-- order they have been inserted into @m@. Unlike @traverse f m@, it does not
+-- fail when there is an element @x@ with @name x /= name (f x)@. It *does* fail
+-- when there are two elements @x@, @y@ with @name (f x) == name (f y)@.
+traverseAsList :: (Applicative f, Named a) => (a -> f a) -> NameMap a -> f (NameMap a)
+traverseAsList f = fmap fromList . Prelude.traverse f . toList
```
Comment (nickchapman-da) [inline libs-haskell/da-hs-base/src/Data/NameMap.hs:141]: Does the runtime check in `traverse` add any real value?

If we were happy to always take the unchecked behaviour (`traverseAsList`) as the behaviour of `traverse`, then could we avoid all the overlapping instance trickiness?

Comment (akrmn) [reply libs-haskell/da-hs-base/src/Data/NameMap.hs:141]: that's a good question. I went with a separate function since the docs for the module state

```haskell
-- 4. Higher-order functions like 'map' and 'traverse' fail when the function
--    passed in tries to change the name of a thing.
```

as one of the features and I thought it would be hard to tell what exactly relies on this. 

It turns out that `NM.traverse` is only used in `DA.Daml.LF.Ast.Optics` (where most changes in this PR take place, and where all uses other than `NameMap {Template,InterfaceCo}Implements` can't possibly change the name) and once in `DA.Daml.LF.Simplifier` (where it's used in a non-name-changing way).

`NM.map` is only used once, also in a non-name-changing way. And now that the `MonoTraversable` instance in `Data.NameMap` is gone and CI passed, we know that no one else was using it. 

tl;dr: I agree with you and I'll use this as the definition of `traverse`, removing promise `4.` from the module. Additionally this would allow `traverse` to change the type of the items, which might be useful later.

Comment (nickchapman-da) [reply libs-haskell/da-hs-base/src/Data/NameMap.hs:141]: > as one of the features and I thought it would be hard to tell what exactly relies on this.

My hand-wavy reasoning was that this run-time check cannot be triggered by the existing code (or else we have a bug that we have never witnessed!)... but your rigorous analysis is far nicer!

(I guess the original purpose of the run-time check was just as a development-time check.)

> tl;dr: I agree with you and I'll use this as the definition of traverse, removing promise 4. from the module. 

:+1:

Comment (nickchapman-da) [reply libs-haskell/da-hs-base/src/Data/NameMap.hs:141]: (I approved in advance of your rework)

Comment (akrmn) [reply libs-haskell/da-hs-base/src/Data/NameMap.hs]: > My hand-wavy reasoning was that this run-time check cannot be triggered by the existing code (or else we have a bug that we have never witnessed!)... but your rigorous analysis is far nicer!

Thank you! Yeah, my worry was that some code might have been written with the expectation that it _should_ fail in such a case, thankfully it wasn't as much code to go through as I thought.

> (I guess the original purpose of the run-time check was just as a development-time check.)

yes that does seem to be the case

> (I approved in advance of your rework)

Thanks! I reverted the `traverseAsList` and type family constraint commits (rather than just rebase them away) so the discussion still makes sense in the future
